astropy/io/fits/tests/test_hdulist.py

```diff
@@ -376,6 +376,24 @@ def test_file_like_3(self):
         info = [(0, 'PRIMARY', 1, 'PrimaryHDU', 5, (100,), 'int32', '')]
         assert fits.info(self.temp('tmpfile.fits'), output=False) == info
 
+    def test_copy_list(self):
+        """
+        Tests that HDUList.copy() returns a shallow copy (regression test
+        for ticket:7211).
+        """
+
+        n = np.arange(10.0)
+        primary_hdu = fits.PrimaryHDU(n)
+        hdu = fits.ImageHDU(n)
+
+        hdul = fits.HDUList([primary_hdu, hdu])
+        new_hdul = hdul.copy()
+        new_hdul[1] = primary_hdu
+
+        assert isinstance(new_hdul, fits.HDUList)
+        assert not new_hdul[1] == hdul[1]
```

Comment:
If you're doing this check, it probably also makes sense to do `assert new_hdul[0] == hdul[0]` as well.

Reply:
Although to be honest I'm not sure if checking for equality between `HDU` objects is that informative anyway.

Reply:
I thought so too. I think a better option would be to use [`id()`](https://docs.python.org/3/library/functions.html#id). How about something like this:
```python
>>> hdul = fits.HDUList([primary_hdu, hdu])
# shallow copy
>>> new_hdul = hdul.copy(deep=False)
# different HDULists
>>> assert not (id(new_hdul) == id(hdul))
# share same item
>>> assert id(new_hdul[0]) == id(hdul[0])
```

Reply:
@ritiek in the case of a shallow copy, it should be possible just to use `is` as the comparator, which should be the same as testing for equivalent `id`s. For the deep copy case, it will be necessary to test object equality, although I'm not whether HDU objects have a meaningful definition of equality.

Reply:
@drdavella Ok, got it. So, this is what we have (not sure about equating HDU objects either):
```python
>>> shallow_copy = hdul.copy(deep=False)
>>> deep_copy = hdul.copy(deep=True)

>>> assert isinstance(shallow_copy, fits.HDUList)
>>> assert isinstance(deep_copy, fits.HDUList)

>>> assert not (id(shallow_copy) == id(hdul))
>>> assert shallow_copy[0] is hdul[0]
>>> assert shallow_copy[1] is hdul[1]

>>> assert not (id(deep_copy) == id(hdul))
>>> assert not deep_copy[0] == hdul[0]
>>> assert not deep_copy[1] == hdul[1]
```
I wanted to know whether this looks good before I update this test downstream (and this conversation becomes hard to follow up :3).

Reply:
@ritiek this is getting there, but it's still not quite right. Here's what I think it should be:
```python
>>> shallow_copy = hdul.copy(deep=False)
>>> deep_copy = hdul.copy(deep=True)

>>> assert isinstance(shallow_copy, fits.HDUList)
>>> assert isinstance(deep_copy, fits.HDUList)

>>> assert shallow_copy is not hdul
>>> assert shallow_copy[0] is hdul[0]
>>> assert shallow_copy[1] is hdul[1]

>>> assert deep_copy is not hdul
>>> assert deep_copy[0] is not hdul[0]
>>> assert deep_copy[1] is not hdul[1]
```

In an ideal world, we'd also be able to do the following:

```python
>>> assert deep_copy[0] == hdul[0]
>>> assert deep_copy[1] == hdul[1]
```

However, since we don't have a good definition of equality for HDUs at the moment, I don't think that will work. It **may** be possible to do this instead:

```python
>>> assert deep_copy[0].header == hdul[0].header
>>> assert deep_copy[1].header == hdul[1].header
>>> (assert deep_copy[0].data == hdul[0].data).all()
>>> (assert deep_copy[1].data == hdul[1].data).all()
```

But that may be overkill.